### PR TITLE
Attribute comment whitespace errors to comment line, not line below it

### LINKED ISSUE #6410: Game throws false error at comment lines
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Endless Sky's data parser throws false errors at comment lines directly above tokens.

### Steps to Reproduce

Copy this and put it in a file. Endless Sky will throw an error that it shouldn't, at the line `planet "Leroo Station". This is caused by the line `   # PLANETS #`, which uses spaces.
```
#############################################################################################
   # PLANETS #
planet "Leroo Station"
	attributes station
	shipyard "Advanced Northern Pirates"
	outfitter "Ammo North"
	outfitter "Basic Outfits"
	outfitter "Basic Weapons"
	outfitter "Pirate Outfits"
	landscape land/stationLeroo
	description "Leroo Station is a pirate station in the Illumina Regions. It serves as some kind of governing building for organized pirate activity. Due to it's size, pirate fleets often use this as a staging point to launch attacks on northwestern traders and occasionally the Deep. The Republic and the Deep remain baffled as to how pirates were able to obtain a large ammount of Bactrians."
	spaceport "At the docking bays, people give you weird looks, most of them being pirates. It's also possible to buy whatever you want for the right price, but this station has no stores selling slaves."
	government Pirate
	"required reputation" 10
	bribe 0.1
	security 0
```

This also triggers a false positive:
```
system Denleitt
	pos -1095.4722 -83.416663
	link Aixafer
	link Littike
	link Luisark
	link Sayreopli
 #
```
Note the last line, which is a comment.

### Expected Behavior

It gets parsed normally.

Endless Sky should NOT ISSUE ERRORS on comment-only lines.

### Screenshots

_No response_

### Link to save file

_No response_

### Operating System

Ubuntu 20.04.3

### Game Source

Built from source

### Game Version

continuous

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by a-random-lemurian**
Fixes #6410.

This branch build will not issue whitespace warnings for these:
```
  # I'm a comment but the parser thinks I'm a whitespace error!
planet "Earth"
    sprite "planet/earth"
    description "Where humans live."
```

Note: Please ignore `6d3c44458e26ab1e22ca15b8f42b03b5bc96d4d5`. That's from a previous pull request and apparently, I suck at git.

### CODE REVIEW

File: source/DataFile.cpp [hunk 1]
```diff
@@ -103,8 +103,11 @@ void DataFile::LoadData(const string &data)
 	vector<int> whiteStack(1, -1);
 	bool fileIsSpaces = false;
 	bool warned = false;
+
+	// Ignore lines which are only comments regardless of whitespace.
+	bool lineIsCommentOnly = false;
 	size_t lineNumber = 0;
-
```
Comment (MCOfficer) [inline source/DataFile.cpp:107]: Unrelated whitespace change

Comment (a-random-lemurian) [reply source/DataFile.cpp:107]: Okay, so what do you think should be done?

Comment (MCOfficer) [reply source/DataFile.cpp:107]: Just restoring that tab should do it. Unfortunately I can't add it as suggestion because i commented on the deleted line. Thanks github :/

Comment (a-random-lemurian) [reply source/DataFile.cpp:107]: I've restored it.

**Comment by Amazinite** [comment]
I don't really see this as a fix. Why is it that we should ignore mixed whitespace on comment lines? For what reason are you using tabs for normal data but spaces for comments (or vice-versa)? If someone used spaces before comments in the code then I'd consider that an issue, so I don't see why it wouldn't be considered an error in the data as well.

**Comment by a-random-lemurian** [comment]
> I don't really see this as a fix. Why is it that we should ignore mixed whitespace on comment lines? For what reason are you using tabs for normal data but spaces for comments (or vice-versa)? If the answer is that you're trying to align a comment, be aware that the size of a tab can differ depending on whatever text editor you're using, so what looks aligned for you might not be aligned for someone else if your tab sizes are different. And if someone used spaces before comments in the code then I'd consider that an issue, so I don't see why it wouldn't be considered an error in the data as well.

Okay, I've decided to change things up a little bit. If this is what you are saying, I see your point. My issue is, Endless Sky attributes the error to the wrong line.
```
# Pretend that the lines below "Earth" are indents
  # I'm a whitespace error, but ES throws an error at the line below me!
planet "Earth"
    sprite "planet/earth"
    description "Where humans live."
```
Endless Sky will attribute the error to `planet "Earth"`, not `  # I'm a comment but the parser thinks I'm a whitespace error!`.

Would this be acceptable?

**Comment by Amazinite** [comment]
Yes.

**Comment by a-random-lemurian** [comment]
I seem to be having trouble, no matter what I do, it doesn't seem to trigger warnings about comment whitespace properly.

Could this be just the behavior of `PrintTrace`?

**Comment by Amazinite** [comment]
The PrintTrace is called from the top of the stack, but comments don't get pushed onto the stack, so we can't actually PrintTrace to the comment. The best I can see us doing is just recognizing that we've hit mixed whitespace on a comment line then having the message say that it's a comment above the line that's shown that's the issue, not the line itself.

**Comment by Amazinite** [comment]
Well it looks like the mixed whitespace warning generally isn't pointing to the right line. Ignoring comments, it's pointing to the line before the whitespace change, not the line with the whitespace change.

**Comment by Amazinite** [comment]
Okay so I think I see what's going on:
* DateFile::LoadData is parsing the entire file into DataNodes.
* New DataNodes are added to the "stack" vector.
* The file is parsed line-by-line.
* If *any* line starts with a space, then the fileIsSpaces bool is set to true.
* If the fileIsSpaces bool is set true and the line starts with a tab, then the "Mixed whitespace usage in file" warning is sent. This means that if you start a line with a space, only the next line that starts with a tab will throw an error.
* The PrintTrace is called *before* the line is tokenized, which means that the mixed whitespace warning is actually being called by the line after the one that is shown.

So take the example that was given:
```
  # I'm a whitespace error, but ES throws an error at the line below me!
planet "Earth"
	sprite "planet/earth"
```
The comment line starts with a space, so fileIsSpaces = true. Then the `planet "Earth"` line is parsed without issue, meaning that that line is now on the top of the stack. Then when `	sprite "planet/earth"` is parsed it sees that the line starts with a tab and that fileIsSpaces is set to true, so it sends the "Mixed whitespace usage in file" warning, but this warning is called on the top of the stack, which means that it's actually printing a trace to the `planet "Earth"` line even though the error is called when the `	sprite "planet/earth"` line is being parsed.

**Comment by Amazinite** [comment]
Working on my own branch to patch up these errors. Here's some example output:
```
Mixed whitespace usage in line
file C:/Users/amaz/Desktop/amazinite-endless-sky/data/categories.txt
L12:   category ship
L18:     "Medium Warship"
L19:       "Heavy Warship"

Mixed whitespace usage for comment at line 24
file C:/Users/amaz/Desktop/amazinite-endless-sky/data/categories.txt
```
The first trace here previously would have stopped at L18, as that's the node that L19 would call a PrintTrace to.
The second trace gives the line number where a comment is using mixed whitespace instead of calling PrintTrace to the line since comments don't get put into DataNodes.

**Comment by a-random-lemurian** [comment]
By the way, will I be credited for discovering this problem in the future 0.9.15 changelog?

**Comment by thomasballinger** [comment]
The example outputs look like they will still match the [regex I wrote for them](https://github.com/thomasballinger/endless-sky-vscode/blob/main/server/src/subprocess.ts#L260-L262), but I'll watch the issue and make sure once the new messages are in.

**Comment by Amazinite** [comment]
Opened #6422 with my changes.

> will I be credited for discovering this problem in the future 0.9.15 changelog?

The way we've done the changelog is that the attribution goes to who fixed the bug.

**Comment by Amazinite** [comment]
Closing in favor of #6422.
